src/awkward/operations/ak_drop_none.py

```diff
@@ -0,0 +1,109 @@
+        <Array [[[0]], [[None]], [[1]], [[2, None]]] type='4 * var * var * ?int64'>
+
+    """
+    with ak._errors.OperationErrorContext(
+        "ak.drop_none",
+        dict(array=array, axis=axis, highlevel=highlevel, behavior=behavior),
+    ):
+        return _impl(array, axis, highlevel, behavior)
+
+
+def _impl(array, axis, highlevel, behavior):
+    layout = ak.operations.to_layout(array, allow_record=False, allow_other=False)
+
+    def drop_nones(layout, **kwargs):
+        if layout.is_list:
+            # only drop nones at list level in the recursion; this way ListArray -> ListOffsetArray with unprojected optiontype -> avoid offset mismatch
+            return layout.drop_none()
+        else:
+            return layout
+
+    if axis is None:
+        # if the outer layout is_option, drop_nones without affecting offsets
+        if layout.is_option:
+            layout = layout.drop_none()
+
+        def action(layout, continuation, **kwargs):
+            return drop_nones(continuation())
+
+    else:
+        max_axis = layout.branch_depth[1] - 1
+        if axis > max_axis:
+            raise ak._errors.wrap_error(
+                np.AxisError(
+                    f"axis={axis} exceeds the depth ({max_axis}) of this array"
+                )
+            )
+
+        def recompute_offsets(layout, depth, depth_context, **kwargs):
+            posaxis = ak._util.maybe_posaxis(layout, axis, depth)
+            if posaxis == 0 and posaxis == depth - 1 and layout.is_list:
+                out = layout._rebuild_without_nones(
+                    options["new_offsets"], layout.content
+                )
+                return out
+            if posaxis == depth and layout.is_list:
+                out = layout._rebuild_without_nones(
+                    options["new_offsets"], layout.content
+                )
+                return out
+
+        def action(layout, depth, depth_context, **kwargs):
+            posaxis = ak._util.maybe_posaxis(layout, axis, depth)
+            if posaxis == 0:
+                if not layout.is_option:
+                    return layout
+                else:
+                    return layout.drop_none()
+            if posaxis == depth - 1 and layout.is_option:
+                _, _, new_offsets = layout._nextcarry_outindex(layout.backend)
+                options.update(new_offsets=new_offsets)
+                return layout.drop_none()
+            if posaxis == depth - 1 and layout.is_list and layout.content.is_option:
+                return layout.drop_none()
+
+    depth_context = {"max_branch_depth": layout.branch_depth[1]}
+    options = {}
+    out = ak._do.recursively_apply(layout, action, behavior, depth_context, options)
+    if "new_offsets" in options and axis is not None:
+        out = ak._do.recursively_apply(
+            out, recompute_offsets, behavior, depth_context, options
+        )
+
+    return ak._util.wrap(out, behavior, highlevel)
```

Comment:
@jpivarski @ioanaif do we want to use `like=` here, or `behavior=ak._util.behaviorof(array, behavior=behavior)`? I can't remember!
```suggestion
    return ak._util.wrap(out, behavior, highlevel, like=behavior)
```